keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,294 @@
+resource version and need to relist (which can impact scalability). See [Watch
+Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
+for details.
+
+### Make the kube-apiserver aware of etcd's minor version
+
+This is only needed if we go with Alternative 2 (Use WatchProgressRequest to
+request watch updates when needed) since Alterative 1 is compatible with all
+etcd 3.x versions.
+
+Allow the etcd minor version to be specified in the kube-apiserver
+`--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  When client
+connections to etcd servers are established (and maybe periodically after that
+as well), check the current etcd version (via the etcd 'version' API) and warn
+the user if the version is older than the version of etcd provided in the
+flag.
+
+if `--storage-backend=etcd3` is provided, the minor etcd version will default to
+the detected etcd version minor version. This is for ease of use, since
+defaulting to `etcd3.0` would require additional configuration management by
+administrators. To make this safe, if the api-server gets an error when making a
+request that is enabled for newer etcd versions, it should warn in the logs that
+there might be an etcd version mismatch and fallback to the etcd functionality
+supported by all etcd 3.x versions.
+
+Also, Due to etcd upgrades and downgrades, there is no way to automatically
+detect the etcd version is a way that is guaranteed to be always correct, so the
+administrator must be able to set the version to a desired version.
+
+In addition to etcd minor version detection, all features requiring features
+introduced at a specific etcd minor version will have feature gates and will go
+through the usual kubernetes stability levels promotions (alpha, beta, GA).
+
+### Risks and Mitigations
+
+Alternative 2 can increase the number of round trips to etcd: One is required to
+get the latest revision, another might be needed to request the progress
+notify. Both of these requests are cheap and scalable, but we need to understand
+the latency impact of waiting for both of them before serving a response back to
+the client. If this turns out to have performance implications, it could be
+partially mitigated by configuring etcd to send regular progress events, which
+is something we will explore.
+
+In the worst case, the “wait until fresh” duration is exceeded, and then the
+request must be served from etcd anyway, which will be higher latency than
+immediately serving from etcd like we do today. We will need to set the
+“wait until fresh” duration to relatively short value (e.g. 20ms), which
+partially mitigates this.
+
+## Design Details
+
+### Pagination
+
+Because the watch cache does not paginate responses, clients that were
+previously getting pagination for RV="" requests will start getting unpaginated
+responses when this feature is enabled.
+
+The impacts of this can be explored when the feature is in alpha.
```

Comment:
I think this doc needs at least a sketch of a plan for dealing with this. Like, this sounds like a showstopper for beta and up, so we need to know that it's solvable or the design is a dead end.

Reply:
I definitely agree with that.

Reply:
Sounds good. I'll explore options and expand this section.